src/WebJobs.Script/Description/DotNet/DotNetFunctionInvoker.cs

```diff
@@ -146,7 +158,7 @@ private void ReloadScript()
                 (_functionSignature == null ||
                 (_functionSignature.HasLocalTypeReference || !_functionSignature.Equals(signature))))
             {
-                Host.RestartEvent.Set();
+                Host.Restart();
```

Comment:
This needs to happen when the function signature changes (as the proxy type that goes down to WebJobs needs to be updated) and when the signature references local types, as those types would now be different and not match the original signature.